File: DecodeImapUtf7Names/ImapUtf7Encoder.pas
unit ImapUtf7Encoder;

interface

//http://www.faqs.org/rfcs/rfc2060.html
//https://www.embarcadero.com/images/dm/technical-papers/delphi-and-unicode-marco-cantu.pdf

type
  TImapUtf7Encoder = class
  private
    class function DecodeBase64Imap(const AValue: string): string;
  public
    class function Decode(const AValue: string): string;
  end;

implementation

uses
  SysUtils, clEncoder, clTranslator;

{ TImapUtf7Encoder }

class function TImapUtf7Encoder.DecodeBase64Imap(const AValue: string): string;
var
  enc: TclEncoder;
  b: TBytes;
begin
  enc := TclEncoder.Create(nil);
  try
    enc.EncodeMethod := cmBase64;
    enc.CharSet := 'UTF-7';
    b := enc.DecodeToBytes(StringReplace(AValue, ',', '/', [rfReplaceAll]));
    Result := TclTranslator.GetString(b, 'UTF-16BE');
  finally
    enc.Free();
  end;
end;

class function TImapUtf7Encoder.Decode(const AValue: string): string;
var
  n, h, t, k: Integer;
begin
  Result := '';
  n := Length(AValue);
  h := 1;

  while (h <= n) do
  begin
    t := System.Pos('&', AValue, h);

    if (t <= 0) then
    begin
      t := n + 1;
    end;

    Result := Result + System.Copy(AValue, h, t - h);
    h := t + 1;
    if (h > n) then Break;

    t := System.Pos('-', AValue, h);
    if (t <= 0) then
    begin
      t := n + 1;
    end;

    k := t - h;
    if (k = 0) then
    begin
      Result := Result + '&';
    end else
    begin
      Result := Result + DecodeBase64Imap(System.Copy(AValue, h, k));
    end;
    h := t + 1;
  end;
end;

end.

File: .net/DecodeImapUtf7NamesNet/ImapUtf7Encoder.cs
using System;
using CleverComponents.InetSuite;

//http://www.faqs.org/rfcs/rfc2060.html

namespace ConsoleApp1
{
	public class ImapUtf7Encoder
	{
		private static string DecodeBase64Imap(string value)
		{
			using (var enc = new Encoder())
			{
				enc.EncodeMethod = EncodeMethod.Base64;

				enc.CharSet = "UTF-7";
				var b = enc.DecodeToBytes(value.Replace(',', '/'));
				return Translator.GetString(b, "UTF-16BE");
			}
		}

		public static string Decode(string value)
		{
			string result = "";
			int n = value.Length;
			int h = 0;

			while (h < n)
			{
				int t = value.IndexOf('&', h);
				if (t < 0)
				{
					t = n;
				}

				result += value.Substring(h, t - h);
				h = t + 1;
				if (h >= n) break;

				t = value.IndexOf('-', h);
				if (t < 0)
				{
					t = n;
				}

				int k = t - h;

				if (k == 0)
				{
					result += '&';
				}
				else
				{
					result += DecodeBase64Imap(value.Substring(h, k));
				}
				h = t + 1;
			}

			return result;
		}
	}
}
